docs/apidiffs/current_vs_latest/opentelemetry-sdk-metrics.txt

```diff
@@ -0,0 +1,266 @@
+	+++  NEW SUPERCLASS: java.lang.Object
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.AggregationTemporality getAggregationTemporality()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) boolean isMonotonic()
++++  NEW INTERFACE: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.SummaryData  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW INTERFACE: io.opentelemetry.sdk.metrics.data.Data
+	+++  NEW SUPERCLASS: java.lang.Object
++++  NEW INTERFACE: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.SummaryPointData  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW INTERFACE: io.opentelemetry.sdk.metrics.data.PointData
+	+++  NEW SUPERCLASS: java.lang.Object
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) long getCount()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) double getSum()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) java.util.List getValues()
++++  NEW INTERFACE: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.ValueAtQuantile  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW SUPERCLASS: java.lang.Object
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) double getQuantile()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) double getValue()
++++  NEW INTERFACE: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.export.MetricExporter  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW INTERFACE: java.io.Closeable
+	+++  NEW INTERFACE: java.lang.AutoCloseable
+	+++  NEW SUPERCLASS: java.lang.Object
+	+++  NEW METHOD: PUBLIC(+) void close()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.common.CompletableResultCode export(java.util.Collection)
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.common.CompletableResultCode flush()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.AggregationTemporality getAggregationTemporality(io.opentelemetry.sdk.metrics.InstrumentType)
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.common.CompletableResultCode shutdown()
++++  NEW INTERFACE: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.export.MetricReader  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW SUPERCLASS: java.lang.Object
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.common.CompletableResultCode flush()
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.metrics.data.AggregationTemporality getAggregationTemporality(io.opentelemetry.sdk.metrics.InstrumentType)
+	+++  NEW METHOD: PUBLIC(+) ABSTRACT(+) io.opentelemetry.sdk.common.CompletableResultCode shutdown()
++++  NEW CLASS: PUBLIC(+) FINAL(+) io.opentelemetry.sdk.metrics.export.PeriodicMetricReader  (not serializable)
+	+++  CLASS FILE FORMAT VERSION: 52.0 <- n.a.
+	+++  NEW INTERFACE: io.opentelemetry.sdk.metrics.export.MetricReader
+	+++  NEW SUPERCLASS: io.opentelemetry.sdk.metrics.internal.export.AbstractMetricReader
```

Comment:
Could we change / remove this superclass without breaking binary compatibility? 🤔

Reply:
I've been trying to adjust the `MetricReader` class / interface structure to try to get rid of this and I'm kind of stuck.

Some observations:
- `MetricReader` has a couple of methods we know we need, like `flush`, `shutdown`, and `getAggregationTemporality`.
- A `MetricReader` also needs some way for a meter provider to register a collection handle to it so that it can trigger collections. We've been calling this collection handle `MetricProducer`. 
- We need to hide the details of a registering a `MetricProducer` because the spec may provide additional options that change the registration contract in the future. 
- We do all that today, with `AbstractMetricReader` hiding the `MetricProducer` registration. 
- The problem is that `AbstractMetricReader` becomes part of the class hierarchy of reader implementations that we need to maintain.

How can we keep `MetricProducer` registration hidden / internal?

Reply:
Ok, might have gotten unstuck. We shouldn't actually need to hide that meter provider needs to register something with metric reader as long as we keep the definition of what that thing is opaque.

So we could extend `MetricReader` with:
```
  void register(SomeEmptyInterface);
```
And then call register on metric reader implementations with an internal class `MetricProducer` that extends `SomeEmptyInterface` with the actual collect method:
```
public interface MetricProducer extends SomeEmptyInterface {
  public Collection<MetricData> collectAllMetrics();
}
```
Metric reader implementations of `register` need to cast `SomeEmptyInterface` to `MetricProducer`.